### PR TITLE
Socket: don't disconnect Socket for unknown/unsupported socket options.

### LINKED ISSUE #59055: GetSocketOption/SetSocketOption should not close underlying socket
### Description

Assume that _socket is an open (active) TCP socket.  The following call will throw an exception **and close the socket**
```csharp
Socket _socket;
_socket.NoDelay = true;
_socket.Blocking = true;
_socket.ReceiveTimeout = -1;
_socket.SendTimeout = -1;
...
IPAddress ipaddr = <setup your server ip>;
IPEndPoint svrEndPoint = new IPEndPoint(ipaddr, _serverPort);
var listener = new Socket(ipaddr.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
listener.Bind(svrEndPoint);
listener.Listen(2);
_socket = listener.Accept();
_socket.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.DontLinger)
```

This option (DontLinger) isn't supported on Linux so the exception is appropriate, but closing the socket isn't.  To quote a friend this "violates the principle of least astonishment".

### Configuration
Version: .NET core 5.0
OS: Linux Ubuntu 20.04.lts
Arch: x64
Configuration: The option isn't supported on this Linux distro but is on Windows.  My code caught the exception but did not expect the socket to be closed.

### Regression?
I do not believe this is a regression.  

### Other information
Here is what I see in the codebase.
in /src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
The GetSocketOption methods call into the OS:
```csharp
...
SocketError errorCode = SocketPal.GetSockOpt(
    _handle,
    optionLevel,
    optionName,
    out optionValue);
...
// Throw an appropriate SocketException if the native call fails.
if (errorCode != SocketError.Success)
{
    UpdateStatusAfterSocketErrorAndThrowException(errorCode);
}
```
So if the call returns a not supported error,  UpdateStatusAfterSocketErrorAndThrowExecption is called.
That calls an internal method named UpdateStatusAfterSocketError which:

```csharp
if (_isConnected && (_handle.IsInvalid || (errorCode != SocketError.WouldBlock &&
        errorCode != SocketError.IOPending && errorCode != SocketError.NoBufferSpaceAvailable &&
        errorCode != SocketError.TimedOut)))
{
    // The socket is no longer a valid socket.
    if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, "Invalidating socket.");
    SetToDisconnected();
}
```
Honestly, that looks correct to me, but perhaps the errorCode isn't what I would be expecting for an unsupported option name, or the handled has somehow (?) become invalid.  I wasn't able to look at this in the debugger, so I can't say what the conditions were or if this is even the right spot (but I don't see anywhere else)

A few nights ago I was down in some mono code and that clearly would have created this effect.  It check the options first and if they failed an exception was thrown (never calling the OS).  I assume that code is no longer used in the v5 .net runtime.

Workaround is the check the Linger option.

### PR DESCRIPTION
**Pull request by tmds**
Fixes https://github.com/dotnet/runtime/issues/59055.

@antonfirsov @dotnet/ncl ptal.

### CODE REVIEW

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 1]
```diff
@@ -3690,30 +3690,38 @@ internal void SetToDisconnected()
...
+            UpdateStatusAfterSocketError(socketException.SocketErrorCode, noDisconnect);
         }
 
-        internal void UpdateStatusAfterSocketError(SocketError errorCode)
+        internal void UpdateStatusAfterSocketError(SocketError errorCode, bool noDisconnect = false)
         {
             // If we already know the socket is disconnected
             // we don't need to do anything else.
-            if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(this, $"errorCode:{errorCode}");
+            if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(this, $"errorCode:{errorCode},noDisconnect:{noDisconnect}");
```
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:3722]:
```suggestion
            if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(this, $"errorCode:{errorCode}, noDisconnect:{noDisconnect}");
```

**Review by antonfirsov — COMMENTED** [review]
Thanks! I have only cosmetic concerns, otherwise looks good.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 2]
```diff
@@ -27,7 +27,7 @@ public void ReuseUnicastPort_CreateSocketGetOption()
                 }
                 else
                 {
-                    Assert.Throws<SocketException>(() => socket.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseUnicastPort));
+                    SocketException se = Assert.Throws<SocketException>(() => socket.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseUnicastPort));
```
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs:30]: Why to assign 'SocketException se' if the variable is unused?

**Comment by tmds** [comment]
@antonfirsov can you take another look? I've addressed your feedback.

I've also made some additional changes. These have to do with us being able to detect use of an unknown option name (as `ENOPROTOOPT`), but not of an unknown option level. For the latter, the `errno` depends on the platform but is ambiguous (e.g. `EINVAL` can also mean the value is not valid).

We no longer disconnect for `ENOPROTOOPT`, but we keep disconnecting for the other errors (so also for unknown levels).

**Review by antonfirsov — APPROVED** [review]
The current changes look good. 

> We no longer disconnect for `ENOPROTOOPT`, but we keep disconnecting for the other errors (so also for unknown levels).

I see you reverted the attempt to fix this. Shall we merge this PR as-is, and  maybe open a new issue to track/discuss possible enhancements?

**Pull request merged**